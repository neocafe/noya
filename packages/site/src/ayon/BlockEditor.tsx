import { useDispatch } from 'noya-app-state-context';
import Sketch from 'noya-file-format';
import {
  FALLTHROUGH,
  getCurrentPlatform,
  handleKeyboardEvent,
} from 'noya-keymap';
import { DrawableLayerType, ParentLayer } from 'noya-state';
import React, {
  ForwardedRef,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
} from 'react';
import {
  BaseEditor,
  createEditor,
  Descendant,
  Editor,
  Node,
  Range,
  Transforms,
} from 'slate';
import { HistoryEditor, withHistory } from 'slate-history';
import { Editable, ReactEditor, Slate, withReact } from 'slate-react';
import styled from 'styled-components';

import {
  allAyonSymbols,
  heading1SymbolId,
  heading2SymbolId,
  heading3SymbolId,
  heading4SymbolId,
  heading5SymbolId,
  heading6SymbolId,
  textSymbolId,
} from './blocks/symbols';
import { filterHashTagsAndSlashCommands } from './parse';
import { Stacking } from './stacking';
import { InferredBlockTypeResult } from './types';
import { useCompletionMenu } from './useCompletionMenu';

export const PositioningElement = styled.div({
  top: '-9999px',
  left: '-9999px',
  position: 'absolute',
  zIndex: Stacking.level.menu,
});

export const ContentElement = styled.div(({ theme }) => ({
  ...theme.textStyles.small,
  borderRadius: 4,
  overflow: 'hidden',
  backgroundColor: theme.colors.popover.background,
  boxShadow: '0 2px 4px rgba(0,0,0,0.2), 0 0 12px rgba(0,0,0,0.1)',
  color: theme.colors.textMuted,
}));

export interface IBlockEditor {
  focus: () => void;
}

const BLOCK_TYPE_SHORTCUTS: { [shortcut: string]: string } = {
  '#': heading1SymbolId,
  '##': heading2SymbolId,
  '###': heading3SymbolId,
  '####': heading4SymbolId,
  '#####': heading5SymbolId,
  '######': heading6SymbolId,
  '"': textSymbolId,
};

function textCommand(
  triggerPrefix: string,
  editor: Editor,
): { range: Range; match: string } | undefined {
  const { selection } = editor;

  if (!selection || !Range.isCollapsed(selection)) return;

  const triggerRegex = new RegExp(`\\${triggerPrefix}(\\w*)$`);

  const [start] = Range.edges(selection);

  const lineStart = Editor.before(editor, start, { unit: 'line' });

  if (!lineStart) return;

  const beforeText = Editor.string(
    editor,
    Editor.range(editor, lineStart, start),
  );

  const beforeMatch = beforeText.match(triggerRegex);

  if (!beforeMatch) return;

  const match = beforeMatch[1];

  const rangeStart = Editor.before(editor, start, {
    unit: 'character',
    distance: match.length + 1,
  });

  if (!rangeStart) return;

  const range = Editor.range(editor, rangeStart, start);

  return { range, match };
}

function textShortcut(
  triggerPrefix: string,
  editor: Editor,
): { range: Range; match: string } | undefined {
  const { selection } = editor;

  if (!selection || !Range.isCollapsed(selection)) return;

  const triggerRegex = new RegExp(`^(${triggerPrefix}) $`);

  const [start] = Range.edges(selection);

  const editorStart = Editor.start(editor, []);

  if (!editorStart) return;

  const beforeText = Editor.string(
    editor,
    Editor.range(editor, editorStart, start),
  );

  const beforeMatch = beforeText.match(triggerRegex);

  if (!beforeMatch) return;

  const match = beforeMatch[1];

  const rangeStart = Editor.before(editor, start, {
    unit: 'character',
    distance: match.length + 1,
  });

  if (!rangeStart) return;

  const range = Editor.range(editor, rangeStart, start);

  return { range, match };
}

export const BlockEditor = forwardRef(function BlockEditor(
  {
    isEditing,
    isSelected,
    layer,
    parent,
    blockText,
    blockTypes,
    onChangeBlockType,
    onFocusCanvas,
  }: {
    isEditing: boolean;
    isSelected: boolean;
    layer: Sketch.SymbolInstance;
    parent: ParentLayer;
    blockText: string;
    blockTypes: InferredBlockTypeResult[];
    onChangeBlockType: (type: DrawableLayerType) => void;
    onFocusCanvas: () => void;
  },
  forwardedRef: ForwardedRef<IBlockEditor>,
) {
  const dispatch = useDispatch();

  const editor = useMemo(() => withHistory(withReact(createEditor())), []);

  const descendants = useMemo(() => fromString(blockText), [blockText]);

  useImperativeHandle(forwardedRef, () => ({
    focus: () => {
      ReactEditor.focus(editor);
      Transforms.select(editor, Editor.range(editor, []));
    },
  }));

  useEffect(() => {
    if (isEditing) return;

    ReactEditor.deselect(editor);
  }, [editor, isEditing]);

  const symbolItems = allAyonSymbols.map((symbol) => ({
    name: symbol.name,
    id: symbol.symbolID,
  }));

  const symbolCompletionMenu = useCompletionMenu({
    editor,
    possibleItems: symbolItems,
    onSelect: (target, item) => {
      Transforms.delete(editor, { at: target });
      const newText = toString(editor.children);

      onChangeBlockType({ symbolId: item.id });
      dispatch('setSymbolIdIsFixed', layer.do_objectID, true);
      dispatch(
        'setBlockText',
        layer.do_objectID,
        newText,
        filterHashTagsAndSlashCommands(newText).content,
      );
    },
  });

  const hashCompletionMenu = useCompletionMenu<{
    name: string;
    id: string;
  }>({
    editor,
    possibleItems: [],
    onSelect: (target, item) => {
      Transforms.delete(editor, { at: target });
      Transforms.insertText(editor, item.id, { at: target.anchor });

      const newText = toString(editor.children);

      dispatch(
        'setBlockText',
        layer.do_objectID,
        newText,
        filterHashTagsAndSlashCommands(newText).content,
      );
    },
  });

  const onKeyDown = useCallback(
    (event: React.KeyboardEvent) => {
      const handleDelete = () => {
        // If there's text, don't delete the layer
        if (blockText) return FALLTHROUGH;

        dispatch('deleteLayer', layer.do_objectID);
        dispatch('interaction', ['reset']);

        onFocusCanvas();
      };

      handleKeyboardEvent(event.nativeEvent, getCurrentPlatform(), {
        Shift: () => {
          dispatch('interaction', ['setCursor', 'cell']);
          return FALLTHROUGH;
        },
        Escape: () => {
          dispatch('interaction', ['reset']);
          dispatch('selectLayer', []);
        },
        Backspace: handleDelete,
        Delete: handleDelete,
        // These may override the Escape shortcut
        ...symbolCompletionMenu.keyMap,
        ...hashCompletionMenu.keyMap,
      });
    },
    [
      symbolCompletionMenu.keyMap,
      hashCompletionMenu.keyMap,
      blockText,
      dispatch,
      layer.do_objectID,
      onFocusCanvas,
    ],
  );

  return (
    <Slate
      editor={editor}
      value={descendants}
      // This can fire after the selection has changed, so we need to be explicit
      // about which layer we're modifying when dispatching actions
      onChange={(value) => {
        const text = toString(value);

        const mdShortcut = textShortcut(
          Object.keys(BLOCK_TYPE_SHORTCUTS).join('|'),
          editor,
        );

        if (mdShortcut) {
          const { range, match } = mdShortcut;

          Transforms.delete(editor, { at: range });
          const newText = toString(editor.children);

          onChangeBlockType({ symbolId: BLOCK_TYPE_SHORTCUTS[match] });
          dispatch('setSymbolIdIsFixed', layer.do_objectID, true);
          dispatch(
            'setBlockText',
            layer.do_objectID,
            newText,
            filterHashTagsAndSlashCommands(newText).content,
          );

          return;
        }

        const slashCommand = textCommand('/', editor);

        if (slashCommand) {
          const { range, match } = slashCommand;

          symbolCompletionMenu.open(range, match);
        } else {
          symbolCompletionMenu.close();
        }

        const hashCommand = textCommand('#', editor);

        if (hashCommand) {
          const { range, match } = hashCommand;

          hashCompletionMenu.open(range, match);
        } else {
          hashCompletionMenu.close();
        }

        dispatch(
          'setBlockText',
          layer.do_objectID,
          text,
          filterHashTagsAndSlashCommands(text).content,
        );

        // Lock the block type when the user starts editing the text
        if (text) {
          dispatch('setSymbolIdIsFixed', layer.do_objectID, true);
        }
      }}
    >
      <Editable
        onKeyDown={onKeyDown}
        style={{ position: 'absolute', inset: 0, padding: 4 }}
      />
      {symbolCompletionMenu.element}
      {hashCompletionMenu.element}
    </Slate>
  );
});

function toString(value: Descendant[]): string {
  return value.map((n) => Node.string(n)).join('\n');
}

function fromString(value: string): Descendant[] {
  return value.split('\n').map((line) => ({
    type: 'paragraph',
    children: [{ text: line }],
  }));
}

export type ParagraphElement = {
  type: 'paragraph';
  align?: string;
  children: Descendant[];
};

type CustomElement = ParagraphElement;

export type CustomText = {
  bold?: boolean;
  italic?: boolean;
  code?: boolean;
  text: string;
};

export type EmptyText = {
  text: string;
};

export type CustomEditor = BaseEditor & ReactEditor & HistoryEditor;

declare module 'slate' {
  interface CustomTypes {
    Editor: CustomEditor;
    Element: CustomElement;
    Text: CustomText | EmptyText;
  }
}
