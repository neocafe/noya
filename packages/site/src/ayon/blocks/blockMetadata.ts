import { Size } from 'noya-geometry';
import {
  avatarSymbolId,
  boxSymbolId,
  buttonSymbolId,
  heroSymbolV2Id,
  iconSymbolId,
  sidebarSymbolId,
  signInSymbolId,
  tableSymbolId,
} from './symbolIds';

export type BlockCategory = 'application' | 'marketing' | 'element';

export type BlockMetadata = {
  name: string;
  category: BlockCategory;
  preferredSize: Size;
};

// This is in a separate file so docs don't have to import Chakra components directly.
// Docs can import this synchronously and import(Ayon)
export const blockMetadata: Record<string, BlockMetadata> = {
  [avatarSymbolId]: {
    name: 'Avatar',
    category: 'element',
    preferredSize: { width: 64, height: 64 },
  },
  [buttonSymbolId]: {
    name: 'Button',
    category: 'element',
    preferredSize: { width: 200, height: 40 },
  },
  [boxSymbolId]: {
    name: 'Box',
    category: 'element',
    preferredSize: { width: 200, height: 200 },
  },
  [heroSymbolV2Id]: {
    name: 'Hero',
    category: 'marketing',
    preferredSize: { width: 600, height: 400 },
  },
  [iconSymbolId]: {
    name: 'Icon',
    category: 'element',
    preferredSize: { width: 48, height: 48 },
  },
  [signInSymbolId]: {
    name: 'Sign In',
    category: 'application',
    preferredSize: { width: 350, height: 250 },
  },
  [tableSymbolId]: {
    name: 'Table',
    category: 'application',
    preferredSize: { width: 600, height: 400 },
  },
  [sidebarSymbolId]: {
    name: 'Sidebar',
    category: 'application',
    preferredSize: { width: 250, height: 600 },
  },
};
