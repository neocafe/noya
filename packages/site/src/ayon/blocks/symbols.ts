import { SketchModel } from 'noya-sketch-model';
import {
  avatarSymbolId,
  boxSymbolId,
  buttonSymbolId,
  checkboxSymbolId,
  headerBarSymbolId,
  heading1SymbolId,
  heading2SymbolId,
  heading3SymbolId,
  heading4SymbolId,
  heading5SymbolId,
  heading6SymbolId,
  heroSymbolId,
  heroSymbolV2Id,
  iconSymbolId,
  imageSymbolId,
  inputSymbolId,
  radioSymbolId,
  selectSymbolId,
  sidebarSymbolId,
  spacerSymbolId,
  switchSymbolId,
  tableSymbolId,
  textareaSymbolId,
  textSymbolId,
  writeSymbolId,
} from './symbolIds';

export const buttonSymbol = SketchModel.symbolMaster({
  symbolID: buttonSymbolId,
  name: 'Button',
});

export const avatarSymbol = SketchModel.symbolMaster({
  symbolID: avatarSymbolId,
  name: 'Avatar',
});

export const boxSymbol = SketchModel.symbolMaster({
  symbolID: boxSymbolId,
  name: 'Box',
});

export const checkboxSymbol = SketchModel.symbolMaster({
  symbolID: checkboxSymbolId,
  name: 'Checkbox',
});

export const inputSymbol = SketchModel.symbolMaster({
  symbolID: inputSymbolId,
  name: 'Input',
});

export const switchSymbol = SketchModel.symbolMaster({
  symbolID: switchSymbolId,
  name: 'Switch',
});

export const textSymbol = SketchModel.symbolMaster({
  symbolID: textSymbolId,
  name: 'Text',
});

export const imageSymbol = SketchModel.symbolMaster({
  symbolID: imageSymbolId,
  name: 'Image',
});

export const heading1Symbol = SketchModel.symbolMaster({
  symbolID: heading1SymbolId,
  name: 'Heading1',
});

export const heading2Symbol = SketchModel.symbolMaster({
  symbolID: heading2SymbolId,
  name: 'Heading2',
});

export const heading3Symbol = SketchModel.symbolMaster({
  symbolID: heading3SymbolId,
  name: 'Heading3',
});

export const heading4Symbol = SketchModel.symbolMaster({
  symbolID: heading4SymbolId,
  name: 'Heading4',
});

export const heading5Symbol = SketchModel.symbolMaster({
  symbolID: heading5SymbolId,
  name: 'Heading5',
});

export const heading6Symbol = SketchModel.symbolMaster({
  symbolID: heading6SymbolId,
  name: 'Heading6',
});

export const writeSymbol = SketchModel.symbolMaster({
  symbolID: writeSymbolId,
  name: 'Write',
});

export const headerBarSymbol = SketchModel.symbolMaster({
  symbolID: headerBarSymbolId,
  name: 'HeaderBar',
});

export const heroSymbol = SketchModel.symbolMaster({
  symbolID: heroSymbolId,
  name: 'HeroLegacyV1',
});

export const heroSymbolV2 = SketchModel.symbolMaster({
  symbolID: heroSymbolV2Id,
  name: 'Hero',
  defaultBlockText:
    '#flex-col #items-center #justify-center #bg-transparent #p-4 #gap-3',
  layers: [
    SketchModel.symbolInstance({
      symbolID: heading2SymbolId,
      blockText: 'Create, iterate, inspire. #text-center',
    }),
    SketchModel.symbolInstance({
      symbolID: textSymbolId,
      blockText: 'Turn great ideas into new possibilities.',
    }),
    SketchModel.symbolInstance({
      symbolID: spacerSymbolId,
      blockText: '#basis-0',
    }),
    SketchModel.symbolInstance({
      symbolID: buttonSymbolId,
      blockText: 'Get started #primary #md',
    }),
  ],
});

export const iconSymbol = SketchModel.symbolMaster({
  symbolID: iconSymbolId,
  name: 'Icon',
});

export const sidebarSymbol = SketchModel.symbolMaster({
  symbolID: sidebarSymbolId,
  name: 'Sidebar',
});

export const tableSymbol = SketchModel.symbolMaster({
  symbolID: tableSymbolId,
  name: 'Table',
});

export const selectSymbol = SketchModel.symbolMaster({
  symbolID: selectSymbolId,
  name: 'Select',
});

export const radioSymbol = SketchModel.symbolMaster({
  symbolID: radioSymbolId,
  name: 'Radio',
});

export const textareaSymbol = SketchModel.symbolMaster({
  symbolID: textareaSymbolId,
  name: 'Textarea',
});

export const spacerSymbol = SketchModel.symbolMaster({
  symbolID: spacerSymbolId,
  name: 'Spacer',
});

export const allAyonSymbols = [
  buttonSymbol,
  avatarSymbol,
  boxSymbol,
  checkboxSymbol,
  inputSymbol,
  switchSymbol,
  textSymbol,
  imageSymbol,
  heading1Symbol,
  heading2Symbol,
  heading3Symbol,
  heading4Symbol,
  heading5Symbol,
  heading6Symbol,
  writeSymbol,
  headerBarSymbol,
  heroSymbol,
  heroSymbolV2,
  iconSymbol,
  sidebarSymbol,
  tableSymbol,
  selectSymbol,
  radioSymbol,
  textareaSymbol,
  // spacerSymbol,
];
