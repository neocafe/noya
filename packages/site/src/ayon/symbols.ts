import { SketchModel } from 'noya-sketch-model';

export const ayonLibraryId = 'e6cc1791-d9c1-410a-84b6-b381fe058bef';

export const buttonSymbolId = 'cdf2346b-cb21-4f23-8d93-7c4fb2e3a5a0';
export const avatarSymbolId = 'fe7dd31d-b140-4e4c-ab7b-90fbe833936b';
export const boxSymbolId = '9dc8f73e-64f2-43f2-8902-af334a7a17cd';
export const checkboxSymbolId = 'fa181f85-04c2-4524-ad4f-779240cb44b0';
export const iconButtonSymbolId = '4343c29a-a05d-421e-9eb6-16365254a5b9';
export const inputSymbolId = '4db59e9d-9bc1-4146-af0a-ec8bf801485f';
export const switchSymbolId = '30a8ac0b-18dd-4744-bee1-77852ec4d3d8';
export const textSymbolId = 'a1f7266f-50cc-416f-9ed2-4af4bca30257';
export const imageSymbolId = 'd91ba1e3-7e64-4966-9cc1-daa48f989178';
export const heading1SymbolId = '955f4cb1-7879-4331-a359-a21a031147cb';
export const heading2SymbolId = 'b3884865-3fbe-4ed0-b577-c3f5864c4a21';
export const heading3SymbolId = 'ab7fb087-28b6-4c9a-a925-22311efb6a19';
export const heading4SymbolId = 'b8441949-12ba-42be-8f87-5cf222af70d0';
export const heading5SymbolId = 'a6f2653e-1595-4dcc-b979-4e0b1340fda3';
export const heading6SymbolId = '467ed35a-0351-4997-ab1e-44b2c6779111';

export const buttonSymbol = SketchModel.symbolMaster({
  symbolID: buttonSymbolId,
  name: 'Button',
});

export const avatarSymbol = SketchModel.symbolMaster({
  symbolID: avatarSymbolId,
  name: 'Avatar',
});

export const boxSymbol = SketchModel.symbolMaster({
  symbolID: boxSymbolId,
  name: 'Box',
});

export const checkboxSymbol = SketchModel.symbolMaster({
  symbolID: checkboxSymbolId,
  name: 'Checkbox',
});

export const iconButtonSymbol = SketchModel.symbolMaster({
  symbolID: iconButtonSymbolId,
  name: 'Icon Button',
});

export const inputSymbol = SketchModel.symbolMaster({
  symbolID: inputSymbolId,
  name: 'Input',
});

export const switchSymbol = SketchModel.symbolMaster({
  symbolID: switchSymbolId,
  name: 'Switch',
});

export const textSymbol = SketchModel.symbolMaster({
  symbolID: textSymbolId,
  name: 'Text',
});

export const imageSymbol = SketchModel.symbolMaster({
  symbolID: imageSymbolId,
  name: 'Image',
});

export const heading1Symbol = SketchModel.symbolMaster({
  symbolID: heading1SymbolId,
  name: 'Heading 1',
});

export const heading2Symbol = SketchModel.symbolMaster({
  symbolID: heading2SymbolId,
  name: 'Heading 2',
});

export const heading3Symbol = SketchModel.symbolMaster({
  symbolID: heading3SymbolId,
  name: 'Heading 3',
});

export const heading4Symbol = SketchModel.symbolMaster({
  symbolID: heading4SymbolId,
  name: 'Heading 4',
});

export const heading5Symbol = SketchModel.symbolMaster({
  symbolID: heading5SymbolId,
  name: 'Heading 5',
});

export const heading6Symbol = SketchModel.symbolMaster({
  symbolID: heading6SymbolId,
  name: 'Heading 6',
});

export const allAyonSymbols = [
  buttonSymbol,
  avatarSymbol,
  boxSymbol,
  checkboxSymbol,
  iconButtonSymbol,
  inputSymbol,
  switchSymbol,
  textSymbol,
  imageSymbol,
  heading1Symbol,
  heading2Symbol,
  heading3Symbol,
  heading4Symbol,
  heading5Symbol,
  heading6Symbol,
];
