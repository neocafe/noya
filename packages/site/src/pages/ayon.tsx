import dynamic from 'next/dynamic';
import React from 'react';

const App = dynamic(() => import('../components/Ayon'), { ssr: false });

export default function Test(): JSX.Element {
  return <App />;
}
