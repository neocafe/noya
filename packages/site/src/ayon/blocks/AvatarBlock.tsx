import { Avatar } from '@chakra-ui/react';
import { BlockDefinition } from 'noya-state';
import React from 'react';
import { isApproximatelySquare, isWithinRectRange } from './score';
import { avatarSymbolId } from './symbols';

export const AvatarBlock: BlockDefinition = {
  id: avatarSymbolId,
  parser: 'regular',
  infer: ({ frame, blockText }) =>
    isWithinRectRange(frame, 30, 30, 120, 120) &&
    isApproximatelySquare(frame, 0.2)
      ? 0.8
      : 0,
  render: (props) => <Avatar size="full" />,
};
