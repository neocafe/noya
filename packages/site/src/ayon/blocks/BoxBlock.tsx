import { Box } from '@chakra-ui/react';
import { BlockDefinition } from 'noya-state';
import React from 'react';
import { filterHashTagsAndSlashCommands } from '../parse';
import { scoreCommandMatch } from './score';
import { boxSymbol, boxSymbolId } from './symbols';
import { isSupportedTailwindClass } from './tailwind';

export const BoxBlock: BlockDefinition = {
  id: boxSymbolId,
  infer: ({ frame, blockText }) =>
    Math.max(scoreCommandMatch(boxSymbol.name, blockText), 0.1),
  render: (props) => {
    const { content, hashTags } = filterHashTagsAndSlashCommands(
      props.blockText,
    );

    const color =
      [content]
        .concat(hashTags)
        .find((value) => CSS.supports('color', `${value}`)) ?? '#ebfdff';

    const hasTailwindBackground = hashTags.some((value) =>
      value.startsWith('bg-'),
    );

    return (
      <Box
        bg={hasTailwindBackground ? undefined : color}
        w="100%"
        h="100%"
        className={hashTags.filter(isSupportedTailwindClass).join(' ')}
      />
    );
  },
};
