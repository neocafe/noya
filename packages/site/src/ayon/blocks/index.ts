import { BlockDefinition } from 'noya-state';
import { AvatarBlock } from './AvatarBlock';
import { BoxBlock } from './BoxBlock';
import { ButtonBlock } from './ButtonBlock';
import { CheckboxBlock } from './CheckboxBlock';
import { HeaderBarBlock } from './HeaderBarBlock';
import {
  Heading1Block,
  Heading2Block,
  Heading3Block,
  Heading4Block,
  Heading5Block,
  Heading6Block,
} from './HeadingBlock';
import { HeroBlock } from './HeroBlock';
import { IconBlock } from './IconBlock';
import { ImageBlock } from './ImageBlock';
import { InputBlock } from './InputBlock';
import { SelectBlock } from './SelectBlock';
import { SidebarBlock } from './SidebarBlock';
import { SwitchBlock } from './SwitchBlock';
import { TableBlock } from './TableBlock';
import { TextBlock } from './TextBlock';
import { WriteBlock } from './WriteBlock';

export const Blocks: Record<string, BlockDefinition> = {
  [AvatarBlock.id]: AvatarBlock,
  [ButtonBlock.id]: ButtonBlock,
  [BoxBlock.id]: BoxBlock,
  [CheckboxBlock.id]: CheckboxBlock,
  [HeaderBarBlock.id]: HeaderBarBlock,
  [Heading1Block.id]: Heading1Block,
  [Heading2Block.id]: Heading2Block,
  [Heading3Block.id]: Heading3Block,
  [Heading4Block.id]: Heading4Block,
  [Heading5Block.id]: Heading5Block,
  [Heading6Block.id]: Heading6Block,
  [HeroBlock.id]: HeroBlock,
  [ImageBlock.id]: ImageBlock,
  [InputBlock.id]: InputBlock,
  [SwitchBlock.id]: SwitchBlock,
  [TextBlock.id]: TextBlock,
  [WriteBlock.id]: WriteBlock,
  [IconBlock.id]: IconBlock,
  [SidebarBlock.id]: SidebarBlock,
  [TableBlock.id]: TableBlock,
  [SelectBlock.id]: SelectBlock,
};
