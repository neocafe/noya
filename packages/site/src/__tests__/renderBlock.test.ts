import * as ChakraUI from '@chakra-ui/react';
import { readFileSync } from 'fs';
import { createElement, createElementCode, format, print } from 'noya-compiler';
import Sketch from 'noya-file-format';
import { SketchModel } from 'noya-sketch-model';
import path from 'path';
import { Blocks } from '../ayon/blocks';
import { buttonSymbolId, textSymbolId } from '../ayon/blocks/symbols';

// Jest doesn't know how to import a text file, so we mock it
jest.mock('../../safelist.txt', () => {
  return {
    default: readFileSync(path.join(__dirname, '../../safelist.txt'), 'utf8'),
  };
});

const Components = new Map<unknown, string>();

Object.entries(ChakraUI).forEach(([key, value]) => {
  Components.set(value, key);
});

function generate(symbol: Sketch.SymbolInstance) {
  return format(
    print(createElementCode(createElement({ Blocks, Components }, symbol)!)),
  );
}

describe('button', () => {
  test('default', () => {
    const symbol = SketchModel.symbolInstance({
      symbolID: buttonSymbolId,
      frame: SketchModel.rect({
        width: 100,
        height: 40,
      }),
    });

    expect(generate(symbol)).toMatchSnapshot();
  });
});

describe('text', () => {
  test('default', () => {
    const symbol = SketchModel.symbolInstance({
      symbolID: textSymbolId,
      blockText: 'Hello world #center #text-red-500',
      frame: SketchModel.rect({
        width: 100,
        height: 40,
      }),
    });

    expect(generate(symbol)).toMatchSnapshot();
  });
});
