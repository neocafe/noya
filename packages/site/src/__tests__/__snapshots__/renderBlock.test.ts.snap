// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button default 1`] = `
"<Frame width={100} height={40}>
  <Button
    size=\\"md\\"
    isFullWidth
    backgroundColor=\\"#f1f5f9\\"
    color=\\"#000\\"
    isDisabled={false}
  >
    {'Submit'}
  </Button>
</Frame>;
"
`;

exports[`text default 1`] = `
"<Frame width={100} height={40}>
  <Text textAlign=\\"center\\" className=\\"text-red-500\\">
    {'Hello world'}
  </Text>
</Frame>;
"
`;
